pkg/ratelimit/runner.go

```diff
@@ -165,34 +188,48 @@ func (cr *ConcurrentRunner) Stop() {
 
 // RunTask runs the task asynchronously.
 func (cr *ConcurrentRunner) RunTask(ctx context.Context, name string, f func(context.Context), opts ...TaskOption) error {
-	taskOpts := &TaskOpts{}
-	for _, opt := range opts {
-		opt(taskOpts)
-	}
 	task := &Task{
-		ctx:  ctx,
-		name: name,
-		f:    f,
-		opts: taskOpts,
+		ctx:      ctx,
+		name:     name,
+		f:        f,
+		priority: constant.Medium,
+	}
+	for _, opt := range opts {
+		opt(task)
 	}
 
 	cr.processPendingTasks()
-	select {
-	case cr.taskChan <- task:
-	default:
-		cr.pendingMu.Lock()
-		defer cr.pendingMu.Unlock()
-		if len(cr.pendingTasks) > 0 {
-			maxWait := time.Since(cr.pendingTasks[0].submittedAt)
-			if maxWait > cr.maxPendingDuration {
-				cr.failedTaskCount.Inc()
-				return ErrMaxWaitingTasksExceeded
-			}
+	cr.pendingMu.Lock()
+	defer cr.pendingMu.Unlock()
+	if task.priority >= constant.High {
+		// We use the max task number to prevent the OOM issue.
+		// It occupies around 1.5GB memory when there is 20000000 pending task.
+		if len(cr.pendingHighPriorityTasks) > maxHighPriorityTaskNum {
+			RunnerFailedTasks.WithLabelValues(cr.name, task.name).Inc()
+			return ErrMaxWaitingTasksExceeded
 		}
 		task.submittedAt = time.Now()
-		cr.pendingTasks = append(cr.pendingTasks, task)
-		cr.pendingTaskCount[task.name]++
+		cr.pendingHighPriorityTasks = append(cr.pendingHighPriorityTasks, task)
```

Comment:
if only one task is submitted, will it always be in the pending queue because you have not sent it to the channel?